python/cog/server/worker.py

```diff
@@ -44,44 +46,164 @@
 class WorkerState(Enum):
     NEW = auto()
     STARTING = auto()
-    READY = auto()
+    IDLE = auto()
     PROCESSING = auto()
+    BUSY = auto()
     DEFUNCT = auto()
 
 
+class Mux:
+    def __init__(self, terminating: asyncio.Event) -> None:
+        self.outs: "defaultdict[str, asyncio.Queue[_PublicEventType]]" = defaultdict(
+            asyncio.Queue
+        )
+        self.terminating = terminating
+        self.fatal: "Optional[FatalWorkerException]" = None
+
+    async def write(self, id: str, item: _PublicEventType) -> None:
+        await self.outs[id].put(item)
+
+    async def read(
+        self, id: str, poll: Optional[float] = None
+    ) -> AsyncIterator[_PublicEventType]:
+        if poll:
+            send_heartbeats = True
+        else:
+            poll = 0.1
+            send_heartbeats = False
+        while 1:
+            try:
+                event = await select(
+                    self.outs[id].get(), self.terminating.wait(), timeout=poll
+                )
+            except TimeoutError:
+                if send_heartbeats:
+                    yield Heartbeat()
+                continue
+            if event is True:
```

Comment:
yeah, I could just use asyncio.wait directly instead of race if you think it'd be better. this seems almost adequate and the type checker can easily verify correctness so I would be inclined to leave it as it is, but maybe it's enough a footgun?